free-2.md

```diff
@@ -1,69 +1,168 @@
+
+
+
 ### Why you?
+
 _What excites you about this idea? How did you come up with it?_
 
 
+
+I've always been interested in teaching/tutoring how people learn, and while I'm not the absolute worst at functional programming languages, learning imperative languages was much easier for me.
+
+
+
 ### Domain
+
 _Describe the project's domain in five words._
 
 
+
+Visualizing general purpose functional languages ?
+
+
+
 ### Interface (syntax)
+
 _How might the user interact with the language? What does programming look 
+
 like? Why is this the right way to interact with the problem domain?_ 
 
 
+
+A program would just be lines of code as per usual.
+
+
+
 ### Operation (semantics)
+
 _What might happen when a program runs? How does a program interact with the
+
 user? What kinds of errors might occur, and how might they be communicated to
+
 the user?_
 
 
+
+The program would update the image when a button is clicked or when the code is executed. If there is an error, it would print the appropriate error message (in whatever style the host language uses), and some problems would be visually represented. For instance, if a function takes in two parameters but only one was given, the user would be able to see that a "piece" is missing; if object types are represented with tabs with the name of the type, then type mismatches would be easy to see.
+
+
+
 ### Expressiveness
+
 _What should be easy to do in this language? What should be possible, but
+
 difficult? What should be impossible or very difficult?_
 
 
+
+It should be easy to write and visualize simple programs. 
+
+
+
 ### Related work
+
 _Are there any other DSLs in this domain? If not, describe how you know there
+
 aren't and conjecture why not. If so, describe them and provide links. How well 
+
 do they address the need? Are there any particularly admirable qualities of the
+
 language? Are there parts of the language you think could be improved?_
 
 
+
+There are a few visualization tools for type inference, but nothing with widespread use. Functional languages just aren't as popular as imperative languages.
+
+
+
 ## The Project
+
 This section examines whether the idea makes for a good CS 111 project.
 
 
+
+I'd argue that it does. It wouldn't be possible to cover everything, but a simple DSL that covers the basics should be possible to make, it's definitely a DSL, and it would help solve the problem.
+
+
+
 ### Suitability
+
 _If someone were to work on this project, what percentage of their time would be
+
 spent directly engaging in the **language** aspects of this project (e.g.,
+
 making language design decisions), as opposed to "systems" aspects of the
+
 project (e.g., implementing a complicated semantics that doesn't require a lot
+
 of language design)?_
 
 
+
+Most of this should involve the language aspects. How to parse the code and create the visualization are the meat of what will be happening, and both of these should be language design decisions.
```

Comment:
Mm, I would say that if you use an existing GPL as the language being coded then parsing it isn't really a design decision.

I agree regarding the visualization though.